File: quickSort/quickSort.py
from math import floor as round_num_down
import random


# quick sort when pivot is selected as median of list
def quick_sort_median_pivot(nums_list=[], left_pointer=0, right_pointer=None):
    if right_pointer is None:
        right_pointer = len(nums_list)-1
    if left_pointer >= right_pointer:
        return nums_list

    def _partition(nums_list, left_pointer, right_pointer):
        pivotIndex = round_num_down((left_pointer+right_pointer)/2)
        pivot = nums_list[pivotIndex]

        while left_pointer <= right_pointer:

            while nums_list[left_pointer] < pivot:
                left_pointer += 1

            while nums_list[right_pointer] > pivot:
                right_pointer -= 1

            if left_pointer <= right_pointer:
                nums_list[left_pointer], nums_list[right_pointer] = nums_list[right_pointer], nums_list[left_pointer]
                left_pointer += 1
                right_pointer -= 1
        return left_pointer

    pivot_index = _partition(nums_list, left_pointer, right_pointer)
    if left_pointer < pivot_index-1:
        quick_sort_median_pivot(nums_list, left_pointer, pivot_index-1)
    if pivot_index < right_pointer:
        quick_sort_median_pivot(nums_list, pivot_index, right_pointer)
    return nums_list


# qick sort when pivot is the first element of list
def quicksort_start_element_pivot(array, start_element=0, end_element=None):
    if end_element is None:
        end_element = len(array) - 1

    def _partition(array, start, end):
        pivot = start
        for i in range(start+1, end+1):
            if array[i] <= array[start]:
                pivot += 1
                array[i], array[pivot] = array[pivot], array[i]
        array[pivot], array[start] = array[start], array[pivot]
        return pivot

    def _quicksort(array, start, end):
        if start >= end:
            return
        pivot = _partition(array, start, end)
        _quicksort(array, start, pivot-1)
        _quicksort(array, pivot+1, end)
        return array

    return _quicksort(array, start_element, end_element)


if __name__ == "__main__":
    LIMIT = 100
    nums_list = [random.randint(0, LIMIT) for _ in range(LIMIT)]

    result_with_median_pivot = quick_sort_median_pivot(nums_list)
    print(result_with_median_pivot)

    result_with_start_element_pivot = quicksort_start_element_pivot(nums_list)
    print(result_with_start_element_pivot)

File: quickSort/quickSort.js
// const generateArr = limit => {
//   const arr = [];
//   while (limit) {
//     arr.push(Math.round(10 - Math.random() * 10));
//     limit--;
//   }
//   return arr;
// };

const _partition = (arr, left, right) => {
  const pivotIndex = Math.floor((left + right) / 2);
  const pivotValue = arr[pivotIndex];

  while (left <= right) {
    while (arr[left] < pivotValue) {
      left++;
    }

    while (arr[right] > pivotValue) {
      right--;
    }

    if (left <= right) {
      const temp = arr[left] + arr[right];
      arr[right] = temp - arr[right];
      arr[left] = temp - arr[right];
      left++;
      right--;
    }
  }
  return left;
};

const quickSortMiddlePivot = (
  arr = [],
  leftPointer = 0,
  rightPointer = null
) => {
  if (!rightPointer) rightPointer = arr.length - 1;

  if (leftPointer >= rightPointer) return arr;

  const pivotIndex = _partition(arr, leftPointer, rightPointer);
  if (leftPointer < pivotIndex - 1) {
    quickSortMiddlePivot(arr, leftPointer, pivotIndex - 1);
  }
  if (pivotIndex < rightPointer) {
    quickSortMiddlePivot(arr, pivotIndex, rightPointer);
  }

  return arr;
};

const main = () => {
  // const LIMIT = 10;
  // const arr = generateArr(LIMIT);
  // console.log(arr);
  const result = quickSortMiddlePivot([4, 3, 6, 1, 7, 2, 9, 0, 8, 5, 0]);
  console.log(result);
};

main();
